### DISCUSSION TITLE
parseAbi meets struct from third lib

### DISCUSSION THREAD

**Discussion by riskers** [wevm/viem#2160] [Question]
How to parse this function by `parseAbi`:

```js
parseAbi([
  'function prepareDeleteObj(address, uint256, uint256, ExtraData memory) external payable returns (uint8, bytes memory, uint256, uint256, address)'
])
```

`ExtraData` is from other file:

```solidity
struct ExtraData {
    address appAddress;
    address refundAddress;
    FailureHandleStrategy failureHandleStrategy;
    bytes callbackData;
}
```

Then `FailureHandleStrategy` is from other file:

```solidity
enum FailureHandleStrategy {
    BlockOnFail,
    CacheOnFail
}
```

`BlockOnFail` / `CacheOnFail` is from `@openzeppelin/contracts-upgradeable/utils/structs/DoubleEndedQueueUpgradeable.sol`.

How to solve it ?

**Comment by tmm** [comment]
You can include struct signatures in `parseAbi`. Enums don't exist in the human-readable ABI spec so you should use `uint8` instead.

[TypeScript Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzmAhlAzgUwIICNhwC+cAZlBCHAOQBuwmIVAUEwMYQB268K+cAvMjRY8wABQBtJnGokArh1YxgnZFEyp1AEUwAbTDEwB5XACsxKACaX16dABo4c4BxgAmAKwA2R89eefOABRAA8YKBQtFBgUOBAGaABPAEo4TDDMKA4UXSFE3n04dRg5LPQ4MT8YAA5HXETDcviQJN8Xd282-064KxtMO2Sqe2lqbig5JWCwiKiYxF7rW3KUMDBsJYH0AG5F-rsizHkOSw39nad26tIUYF1SzAAJFBP9AGVw6MwAc0Td+sacFYOV0uBQrAA1nMULtCMMmABdZJMXjAJgAenRcAAegB+JhAA)

A good way to reverse engineer the arguments to pass to `parseAbi` is to use [`formatAbi`](https://abitype.dev/api/human#formatabi-1) to convert a JSON ABI to human-readable.
